packages/animations/lib/src/dual_transition_builder.dart

```diff
@@ -0,0 +1,228 @@
+  BuildContext context,
+  Animation<double> animation,
+  Widget child,
+);
+
+/// A animated builder that animates its [child] with different transitions
+/// based on the [AnimationStatus] of the provided [animation].
+///
+/// While the [animation] runs forward, the [child] is animated according to
+/// [forwardBuilder] and while the [animation] is running in reverse, it is
+/// animated according to [reverseBuilder].
+///
+/// This widget can be used to specify different enter and exit transitions for
+/// a [child].
+class DualTransitionBuilder extends StatefulWidget {
+  /// Creates a [DualTransitionBuilder].
+  ///
+  /// The [animation], [forwardBuilder], and [reverseBuilder] arguments are
+  /// required and must not be null.
+  const DualTransitionBuilder({
+    Key key,
+    @required this.animation,
+    @required this.forwardBuilder,
+    @required this.reverseBuilder,
+    this.child,
+  })  : assert(animation != null),
+        assert(forwardBuilder != null),
+        assert(reverseBuilder != null),
+        super(key: key);
+
+  /// The animation that drives the [child]'s transition.
+  ///
+  /// When this animation runs forward, the [child] transitions as specified by
+  /// [forwardBuilder]. When it runs in reverse, the child transitions according
+  /// to [reverseBuilder].
+  final Animation<double> animation;
+
+  /// A builder for a transition that wraps [child].
+  ///
+  /// The `animation` provided to this builder is running forward from 0.0 to
+  /// 1.0 when [animation] runs _forward_. When [animation] runs in reverse,
+  /// the given animation is set to [kAlwaysCompleteAnimation].
+  ///
+  /// This builder should be used to create the animation that makes the [child]
+  /// appear on screen. The [child] should be fully visible when the `animation`
+  /// reaches 1.0.
+  ///
+  /// The builder will be called whenever the [Animation.value] of the given
+  /// `animation` changes.
+  ///
+  /// See also:
+  ///
+  ///  * [reverseBuilder], which builds the transition for making the [child]
+  ///   disappear from the screen.
+  final TransitionBuilder forwardBuilder;
+
+  /// A builder for a transition that wraps [child].
+  ///
+  /// The `animation` provided to this builder is running forward from 0.0 to
+  /// 1.0 when [animation] runs in _reverse_. When [animation] runs forward,
+  /// the given animation is set to [kAlwaysDismissedAnimation].
+  ///
+  /// This builder should be used to create the animation that makes the [child]
+  /// disappear from the screen. The [child] should be fully invisible when the
+  /// `animation` reaches 1.0.
+  ///
+  /// The builder will be called whenever the [Animation.value] of the given
+  /// `animation` changes.
+  ///
+  /// See also:
+  ///
+  ///  * [forwardBuilder], which builds the transition for making the [child]
+  ///    appear on screen.
+  final TransitionBuilder reverseBuilder;
+
+  /// The widget below this widget in the tree.
```

Comment:
To prevent confusion from "the widget" and "this widget"
```suggestion
  /// The widget below this [DualTransitionBuilder] in the tree.
```